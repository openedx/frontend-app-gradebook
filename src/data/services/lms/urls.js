import { getConfig } from '@edx/frontend-platform';
import { StrictDict } from 'utils';
import { historyRecordLimit } from './constants';
import { filterQuery, stringifyUrl } from './utils';

const courseId = window.location.pathname.split('/').filter(Boolean).pop() || '';

export const getUrlPrefix = () => `${getConfig().LMS_BASE_URL}/api/`;
export const getBulkGradesUrl = () => `${getUrlPrefix()}bulk_grades/course/${courseId}/`;
export const getEnrollmentUrl = () => `${getUrlPrefix()}enrollment/v1/`;
export const getGradesUrl = () => `${getUrlPrefix()}grades/v1/`;
export const getGradebookUrl = () => `${getGradesUrl()}gradebook/${courseId}/`;
export const getBulkUpdateUrl = () => `${getGradebookUrl()}bulk-update`;
export const getInterventionUrl = () => `${getBulkGradesUrl()}intervention/`;
export const getCohortsUrl = () => `${getUrlPrefix()}courses/${courseId}/cohorts/`;
export const getTracksUrl = () => `${getEnrollmentUrl()}course/${courseId}?include_expired=1`;
export const getBulkHistoryUrl = () => `${getBulkUpdateUrl()}history/`;
export const getAssignmentTypesUrl = () => stringifyUrl(`${getGradebookUrl()}grading-info`, { graded_only: true });
export const getRolesUrl = () => stringifyUrl(`${getEnrollmentUrl()}roles/`, { courseId });
/**
 * bulkGradesUrlByCourseAndRow(courseId, rowId)
 * returns the bulkGrades url with the given rowId.
 * @param {string} rowId - row/error identifier
 * @return {string} - bulk grades fetch url
 */
export const bulkGradesUrlByRow = (rowId) => stringifyUrl(getBulkGradesUrl(), { error_id: rowId });

export const gradeCsvUrl = (options = {}) => stringifyUrl(getBulkGradesUrl(), filterQuery(options));

export const interventionExportCsvUrl = (options = {}) => (
  stringifyUrl(getInterventionUrl(), filterQuery(options))
);

export const sectionOverrideHistoryUrl = (subsectionId, userId) => stringifyUrl(
  `${getGradesUrl()}subsection/${subsectionId}/`,
  { user_id: userId, history_record_limit: historyRecordLimit },
);

export default StrictDict({
  getUrlPrefix,
  getBulkGradesUrl,
  getEnrollmentUrl,
  getGradesUrl,
  getGradebookUrl,
  getBulkUpdateUrl,
  getInterventionUrl,
  getCohortsUrl,
  getTracksUrl,
  getBulkHistoryUrl,
  getAssignmentTypesUrl,
  getRolesUrl,
  bulkGradesUrlByRow,
  gradeCsvUrl,
  interventionExportCsvUrl,
  sectionOverrideHistoryUrl,
});
