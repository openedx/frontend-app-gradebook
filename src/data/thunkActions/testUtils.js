/* eslint-disable import/no-extraneous-dependencies */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

/** createTestFetcher(mockedMethod, thunkAction, args, onDispatch)
 * Creates a testFetch method, which will test a given thunkAction of the form:
 *   ```
 *   const <thunkAction> = (<args>) => (dispatch, getState) => {
 *   ...
 *   return <mockedMethod>.then().catch();
 *   ```
 * The returned function will take a promise handler function, a list of expected actions
 * to have been dispatched (objects only), and an optional verifyFn method to be called after
 * the fetch has been completed.
 *
 * @param {fn} mockedMethod - already-mocked api method being exercised by the thunkAction.
 * @param {fn} thunkAction - thunkAction to call/test
 * @param {array} args - array of args to dispatch the thunkAction with
 * @param {[fn]} onDispatch - optional function to be called after dispatch
 *
 * @return {fn} testFetch method
 *   @param {fn} resolveFn - promise handler of the form (resolve, reject) => {}.
 *     should return a call to resolve or reject with response data.
 *   @param {object[]} expectedActions - array of action objects expected to have been dispatched
 *     will be verified after the thunkAction resolves
 *   @param {[fn]} verifyFn - optional function to be called after dispatch
 */
export const createTestFetcher = (
  mockedMethod,
  thunkAction,
  args,
  onDispatch,
) => (
  resolveFn,
  expectedActions,
) => {
  const store = mockStore({});
  mockedMethod.mockReturnValue(new Promise(resolve => {
    resolve(new Promise(resolveFn));
  }));
  return store.dispatch(thunkAction(...args)).then(() => {
    onDispatch();
    if (expectedActions !== undefined) {
      expect(store.getActions()).toEqual(expectedActions);
    }
  });
};

export default {
  createTestFetcher,
};
