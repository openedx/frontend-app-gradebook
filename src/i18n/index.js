import { messages as footerMessages } from '@edx/frontend-component-footer';
import { messages as headerMessages } from '@edx/frontend-component-header';
import { messages as paragonMessages } from '@edx/paragon';

import arMessages from './messages/ar.json';
import deMessages from './messages/de.json';
import es419Messages from './messages/es_419.json';
import faIRMessages from './messages/fa_IR.json';
import frMessages from './messages/fr.json';
import frCAMessages from './messages/fr_CA.json';
import hiMessages from './messages/hi.json';
import itMessages from './messages/it.json';
import ptMessages from './messages/pt.json';
import ruMessages from './messages/ru.json';
import ukMessages from './messages/uk.json';
import zhcnMessages from './messages/zh_CN.json';
// no need to import en messages-- they are in the defaultMessage field

const appMessages = {
  ar: arMessages,
  'es-419': es419Messages,
  'fa-ir': faIRMessages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  pt: ptMessages,
  it: itMessages,
  de: deMessages,
  hi: hiMessages,
  'fr-ca': frCAMessages,
  ru: ruMessages,
  uk: ukMessages,
};

export default [
  paragonMessages,
  footerMessages,
  headerMessages,
  appMessages,
];
