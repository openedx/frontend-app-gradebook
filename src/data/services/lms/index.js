import { StrictDict } from '@src/utils';
import api from './api';
import urls from './urls';

export default StrictDict({
  api,
  urls,
});
