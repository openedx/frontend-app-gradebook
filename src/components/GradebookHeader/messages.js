import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  backToDashboard: {
    id: 'gradebook.GradebookHeader.backButton',
    defaultMessage: 'Back to Dashboard',
    description: 'Button text to take user back to LMS dashboard in Gradebook Header',
  },
  gradebook: {
    id: 'gradebook.GradebookHeader.appLabel',
    defaultMessage: 'Gradebook',
    description: 'Top-level app title in Gradebook Header component',
  },
  frozenWarning: {
    id: 'gradebook.GradebookHeader.frozenWarning',
    defaultMessage: 'The grades for this course are now frozen. Editing of grades is no longer allowed.',
    description: 'Warning message in Gradebook Header for frozen messages',
  },
  unauthorizedWarning: {
    id: 'gradebook.GradebookHeader.unauthorizedWarning',
    defaultMessage: 'You are not authorized to view the gradebook for this course.',
    description: 'Warning message in Gradebook Header when user is not allowed to view the app',
  },
  toActivityLog: {
    id: 'gradebook.GradebookHeader.toActivityLogButton',
    defaultMessage: 'View Bulk Management History',
    description: 'Button text for button navigating to Bulk Managment Activity Log',
  },
  toGradesView: {
    id: 'gradebook.GradebookHeader.toGradesView',
    defaultMessage: 'Return to Gradebook',
    description: 'Button text for button navigating to Grades view.',
  },
});

export default messages;
