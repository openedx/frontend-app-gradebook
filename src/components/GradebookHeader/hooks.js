import { views } from 'data/constants/app';
import { actions, selectors } from 'data/redux/hooks';

import messages from './messages';

export const useGradebookHeaderData = () => {
  const activeView = selectors.app.useActiveView();
  const courseId = selectors.app.useCourseId();
  const areGradesFrozen = selectors.assignmentTypes.useAreGradesFrozen();
  const canUserViewGradebook = selectors.roles.useCanUserViewGradebook();
  const showBulkManagement = selectors.root.useShowBulkManagement();
  const setView = actions.app.useSetView();

  const handleToggleViewClick = () => setView(
    activeView === views.grades
      ? views.bulkManagementHistory
      : views.grades,
  );

  const toggleViewMessage = activeView === views.grades
    ? messages.toActivityLog
    : messages.toGradesView;

  return {
    areGradesFrozen,
    canUserViewGradebook,
    courseId,
    showBulkManagement,

    handleToggleViewClick,
    toggleViewMessage,
  };
};

export default useGradebookHeaderData;
