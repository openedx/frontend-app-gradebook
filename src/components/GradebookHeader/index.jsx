import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';

import { instructorDashboardUrl } from 'data/services/lms/urls';
import useGradebookHeaderData from './hooks';
import messages from './messages';

export const GradebookHeader = () => {
  const { formatMessage } = useIntl();
  const {
    areGradesFrozen,
    canUserViewGradebook,
    courseId,
    handleToggleViewClick,
    showBulkManagement,
    toggleViewMessage,
  } = useGradebookHeaderData();
  const dashboardUrl = instructorDashboardUrl();
  return (
    <div className="gradebook-header">
      <a href={dashboardUrl} className="mb-3">
        <span aria-hidden="true">{'<< '}</span>
        {formatMessage(messages.backToDashboard)}
      </a>
      <h1>{formatMessage(messages.gradebook)}</h1>
      <div className="subtitle-row d-flex justify-content-between align-items-center">
        <h2>{courseId}</h2>
        {showBulkManagement && (
          <Button variant="tertiary" onClick={handleToggleViewClick}>
            {formatMessage(toggleViewMessage)}
          </Button>
        )}
      </div>
      {areGradesFrozen && (
        <div className="alert alert-warning" role="alert">
          {formatMessage(messages.frozenWarning)}
        </div>
      )}
      {(canUserViewGradebook === false) && (
        <div className="alert alert-warning" role="alert">
          {formatMessage(messages.unauthorizedWarning)}
        </div>
      )}
    </div>
  );
};

export default GradebookHeader;
