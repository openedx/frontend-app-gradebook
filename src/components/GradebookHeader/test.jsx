import React from 'react';
import { shallow } from 'enzyme';

import { Button } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import selectors from 'data/selectors';
import { views } from 'data/constants/app';
import messages from './messages';
import { GradebookHeader, mapDispatchToProps, mapStateToProps } from '.';

jest.mock('@edx/paragon', () => ({
  Button: () => 'Button',
}));
jest.mock('@edx/frontend-platform/i18n', () => ({
  defineMessages: m => m,
  FormattedMessage: () => 'FormattedMessage',
}));
jest.mock('data/actions', () => ({
  __esModule: true,
  default: {
    app: { setView: jest.fn() },
  },
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      activeView: jest.fn(state => ({ aciveView: state })),
      courseId: jest.fn(state => ({ courseId: state })),
    },
    assignmentTypes: { areGradesFrozen: jest.fn(state => ({ areGradesFrozen: state })) },
    roles: { canUserViewGradebook: jest.fn(state => ({ canUserViewGradebook: state })) },
    root: { showBulkManagement: jest.fn(state => ({ showBulkManagement: state })) },
  },
}));

const courseId = 'fakeID';
describe('GradebookHeader component', () => {
  const props = {
    activeView: views.grades,
    areGradesFrozen: false,
    canUserViewGradebook: false,
    courseId,
    showBulkManagement: false,
  };
  beforeEach(() => {
    props.setView = jest.fn();
  });
  describe('snapshots', () => {
    let el;
    beforeEach(() => {
      el = shallow(<GradebookHeader {...props} />);
      el.instance().handleToggleViewClick = jest.fn().mockName('this.handleToggleViewClick');
    });
    describe('default values (grades frozen, cannot view).', () => {
      test('unauthorized warning, but no grades frozen warning', () => {
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('grades frozen, cannot view', () => {
      test('unauthorized warning, and grades frozen warning.', () => {
        el.setProps({ areGradesFrozen: true });
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('grades frozen, can view.', () => {
      test('grades frozen warning but no unauthorized warning', () => {
        el.setProps({ areGradesFrozen: true, canUserViewGradebook: true });
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('show bulk management, active view is grades view', () => {
      test('toggle view button to activity log', () => {
        el.setProps({ showBulkManagement: true });
        expect(el.find(Button).getElement()).toEqual((
          <Button
            variant="tertiary"
            onClick={el.instance().handleToggleViewClick}
          >
            <FormattedMessage {...messages.toActivityLog} />
          </Button>
        ));
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
    describe('show bulk management, active view is bulkManagementHistory view', () => {
      test('toggle view button to grades', () => {
        el.setProps({ showBulkManagement: true, activeView: views.bulkManagementHistory });
        expect(el.find(Button).getElement()).toEqual((
          <Button
            variant="tertiary"
            onClick={el.instance().handleToggleViewClick}
          >
            <FormattedMessage {...messages.toGradesView} />
          </Button>
        ));
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
  });
  describe('behavior', () => {
    let el;
    beforeEach(() => {
      el = shallow(<GradebookHeader {...props} />);
    });
    describe('handleToggleViewClick', () => {
      test('calls setView with activity view if activeView is grades', () => {
        el.instance().handleToggleViewClick();
        expect(props.setView).toHaveBeenCalledWith(views.bulkManagementHistory);
      });
      test('calls setView with grades view if activeView is bulkManagementHistory', () => {
        el.setProps({ activeView: views.bulkManagementHistory });
        el.instance().handleToggleViewClick();
        expect(props.setView).toHaveBeenCalledWith(views.grades);
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { a: 'test', example: 'state' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('activeView from app.activeView', () => {
      expect(mapped.activeView).toEqual(selectors.app.activeView(testState));
    });
    test('courseId from app.courseId', () => {
      expect(mapped.courseId).toEqual(selectors.app.courseId(testState));
    });
    test('areGradesFrozen from assignmentTypes selector', () => {
      expect(
        mapped.areGradesFrozen,
      ).toEqual(selectors.assignmentTypes.areGradesFrozen(testState));
    });
    test('canUserViewGradebook from roles selector', () => {
      expect(
        mapped.canUserViewGradebook,
      ).toEqual(selectors.roles.canUserViewGradebook(testState));
    });
    test('showBulkManagement from root showBulkManagement selector', () => {
      expect(mapped.showBulkManagement).toEqual(selectors.root.showBulkManagement(testState));
    });
  });

  describe('mapDispatchToProps', () => {
    test('setView from actions.app.setView', () => {
      expect(mapDispatchToProps.setView).toEqual(actions.app.setView);
    });
  });
});
