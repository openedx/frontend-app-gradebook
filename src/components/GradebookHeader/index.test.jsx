import React from 'react';
import { shallow } from 'enzyme';
import { useIntl } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';

import { formatMessage } from 'testUtils';
import { instructorDashboardUrl } from 'data/services/lms/urls';

import useGradebookHeaderData from './hooks';
import GradebookHeader from '.';

jest.mock('./hooks', () => ({ __esModule: true, default: jest.fn() }));
jest.mock('data/services/lms/urls', () => ({
  instructorDashboardUrl: jest.fn(),
}));

instructorDashboardUrl.mockReturnValue('test-dashboard-url');

const hookProps = {
  areGradesFrozen: false,
  canUserViewGradebook: true,
  courseId: 'test-course-id',
  handleToggleViewClick: jest.fn().mockName('hooks.handleToggleViewClick'),
  showBulkManagement: false,
  toggleViewMessage: { defaultMessage: 'toggle-view-message' },
};
useGradebookHeaderData.mockReturnValue(hookProps);

let el;
describe('GradebookHeader component', () => {
  beforeAll(() => {
    el = shallow(<GradebookHeader />);
  });
  describe('behavior', () => {
    it('initializes hooks', () => {
      expect(useGradebookHeaderData).toHaveBeenCalledWith();
      expect(useIntl).toHaveBeenCalledWith();
    });
  });
  describe('render', () => {
    describe('default view', () => {
      test('shapshot', () => {
        expect(el).toMatchSnapshot();
      });
    });
    describe('show bulk management', () => {
      beforeEach(() => {
        useGradebookHeaderData.mockReturnValueOnce({ ...hookProps, showBulkManagement: true });
        el = shallow(<GradebookHeader />);
      });
      test('snapshot: show toggle view message button with handleToggleViewClick method', () => {
        expect(el).toMatchSnapshot();
        const { onClick, children } = el.find(Button).props();
        expect(onClick).toEqual(hookProps.handleToggleViewClick);
        expect(children).toEqual(formatMessage(hookProps.toggleViewMessage));
      });
    });
    describe('frozen grades', () => {
      beforeEach(() => {
        useGradebookHeaderData.mockReturnValueOnce({ ...hookProps, areGradesFrozen: true });
        el = shallow(<GradebookHeader />);
      });
      test('snapshot: show frozen warning', () => {
        expect(el).toMatchSnapshot();
      });
    });
    describe('user cannot view gradebook', () => {
      beforeEach(() => {
        useGradebookHeaderData.mockReturnValueOnce({ ...hookProps, canUserViewGradebook: false });
        el = shallow(<GradebookHeader />);
      });
      test('snapshot: show unauthorized warning', () => {
        expect(el).toMatchSnapshot();
      });
    });
  });
});
