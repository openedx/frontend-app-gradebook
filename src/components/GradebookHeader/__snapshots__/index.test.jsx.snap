// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradebookHeader component render default view shapshot 1`] = `
<div
  className="gradebook-header"
>
  <a
    className="mb-3"
    href="test-dashboard-url"
  >
    <span
      aria-hidden="true"
    >
      &lt;&lt; 
    </span>
    Back to Dashboard
  </a>
  <h1>
    Gradebook
  </h1>
  <div
    className="subtitle-row d-flex justify-content-between align-items-center"
  >
    <h2>
      test-course-id
    </h2>
  </div>
</div>
`;

exports[`GradebookHeader component render frozen grades snapshot: show frozen warning 1`] = `
<div
  className="gradebook-header"
>
  <a
    className="mb-3"
    href="test-dashboard-url"
  >
    <span
      aria-hidden="true"
    >
      &lt;&lt; 
    </span>
    Back to Dashboard
  </a>
  <h1>
    Gradebook
  </h1>
  <div
    className="subtitle-row d-flex justify-content-between align-items-center"
  >
    <h2>
      test-course-id
    </h2>
  </div>
  <div
    className="alert alert-warning"
    role="alert"
  >
    The grades for this course are now frozen. Editing of grades is no longer allowed.
  </div>
</div>
`;

exports[`GradebookHeader component render show bulk management snapshot: show toggle view message button with handleToggleViewClick method 1`] = `
<div
  className="gradebook-header"
>
  <a
    className="mb-3"
    href="test-dashboard-url"
  >
    <span
      aria-hidden="true"
    >
      &lt;&lt; 
    </span>
    Back to Dashboard
  </a>
  <h1>
    Gradebook
  </h1>
  <div
    className="subtitle-row d-flex justify-content-between align-items-center"
  >
    <h2>
      test-course-id
    </h2>
    <Button
      onClick={[MockFunction hooks.handleToggleViewClick]}
      variant="tertiary"
    >
      toggle-view-message
    </Button>
  </div>
</div>
`;

exports[`GradebookHeader component render user cannot view gradebook snapshot: show unauthorized warning 1`] = `
<div
  className="gradebook-header"
>
  <a
    className="mb-3"
    href="test-dashboard-url"
  >
    <span
      aria-hidden="true"
    >
      &lt;&lt; 
    </span>
    Back to Dashboard
  </a>
  <h1>
    Gradebook
  </h1>
  <div
    className="subtitle-row d-flex justify-content-between align-items-center"
  >
    <h2>
      test-course-id
    </h2>
  </div>
  <div
    className="alert alert-warning"
    role="alert"
  >
    You are not authorized to view the gradebook for this course.
  </div>
</div>
`;
